daemon/src/device.rs

```diff
@@ -74,6 +78,57 @@ struct ButtonState {
     hold_handled: bool,
 }
 
+#[derive(Debug, Clone, Default)]
+struct TapTempoState {
+    taps: Vec<Instant>,
+}
+
+impl TapTempoState {
+    fn reset_start(&mut self, now: Instant) {
+        self.taps.clear();
+        self.taps.push(now);
+    }
+
+    fn last(&self) -> Option<Instant> {
+        self.taps.last().copied()
+    }
+
+    fn register(&mut self, now: Instant, window: Duration) -> Option<Vec<Duration>> {
+        // If this is the first tap or we've exceeded the window, start fresh
+        if self.taps.is_empty() {
+            self.reset_start(now);
+            return None;
+        }
+
+        if let Some(last) = self.last() {
+            if now.duration_since(last) > window {
+                self.reset_start(now);
+                return None;
+            }
+        }
+
+        // Add the new tap
+        self.taps.push(now);
+
+        // Keep the vector small (we only need a few recent taps)
+        const TAP_TEMPO_MAX_TAPS: usize = 10;
+        if self.taps.len() > TAP_TEMPO_MAX_TAPS {
+            let overflow = self.taps.len() - TAP_TEMPO_MAX_TAPS;
+            self.taps.drain(0..overflow);
+        }
+
+        // Compute durations using all recent intervals once we have at least 4 taps
+        if self.taps.len() >= 4 {
```

Comment:
Why do we need a minimum of 4 taps? It should be possible to work out the tempo with a minimum of 2 taps on the button, extra taps are useful for refining the granularity.